### PR TITLE
Uses a better body reader for the http request of the grpc server.

### PR DESCRIPTION
**Pull request by cyriltovena**
It might not be obvious but this small changes has a big impact:

1. if you read the `http.NewRequest` code you will realize that the content lenght is computed only if the body interface is from a specific type.

```go
if body != nil {
		switch v := body.(type) {
		case *bytes.Buffer:
			req.ContentLength = int64(v.Len())
			buf := v.Bytes()
			req.GetBody = func() (io.ReadCloser, error) {
				r := bytes.NewReader(buf)
				return ioutil.NopCloser(r), nil
			}
		case *bytes.Reader:
			req.ContentLength = int64(v.Len())
			snapshot := *v
			req.GetBody = func() (io.ReadCloser, error) {
				r := snapshot
				return ioutil.NopCloser(&r), nil
			}
		case *strings.Reader:
			req.ContentLength = int64(v.Len())
			snapshot := *v
			req.GetBody = func() (io.ReadCloser, error) {
				r := snapshot
				return ioutil.NopCloser(&r), nil
			}
		default:
			// This is where we'd set it to -1 (at least
			// if body != NoBody) to mean unknown, but
			// that broke people during the Go 1.8 testing
			// period. People depend on it being 0 I
			// guess. Maybe retry later. See Issue 18117.
		}

```

This means that until then, the `ContentLength` was not available to downstream handler.

2. I could have simply remove the useless `ioutil.NopCloser` but instead I replace it with `bytes.Buffer`. Now this means we don't need anymore to make a copy of the body if we want to use it. We can just cast it.

For example in Cortex this code https://github.com/cortexproject/cortex/blob/master/pkg/util/http.go#L94-L99 will never allocate properly the buffer and will make multiple allocations call while reading the body because `expectedSize` is always 0.

This happens in Loki too, and most likely in Tempo.

Signed-off-by: Cyril Tovena <cyril.tovena@gmail.com>

### CODE REVIEW

**Comment by cyriltovena** [comment]
I've found an even better way. NewRequest does the wrapping of ioutil.NopCloser if not done, and that makes it impossible to cast back to a buffer.

**Review by pracucci — COMMENTED** [review]
Interesting! Changes LGTM.

**Comment by bboreham** [comment]
I found it rather hard to follow this code, because it's really doing two different things, and the second commit undoes part of the first commit without explanation. I think I understand what is intended now, after some Slack conversation.

IMHO these should be two PRs, or at least two non-overlapping commits.

The addition of `nopCloser` should be motivated; you declare a similar thing to be "useless" in your description.

Also I think the naming of `BodyFromReader` could better describe what (I think) you are doing - providing fast access to someone who wants a `[]byte` from a `Reader`.  And either explain why the parameter is `ReadCloser` or change it.

File: httpgrpc/server/server.go [hunk 1]
```diff
@@ -37,16 +38,35 @@ func NewServer(handler http.Handler) *Server {
...
+}
+
+func (nopCloser) Close() error { return nil }
+
+func nopCloserBuffer(b *bytes.Buffer) io.ReadCloser {
+	return nopCloser{b}
+}
+
+// BodyFromReader attempt to return the body as a bytes buffer to avoid reading and copying one that already exists.
+func BodyFromReader(r io.ReadCloser) (*bytes.Buffer, bool) {
```
Comment (bboreham) [inline httpgrpc/server/server.go:52]: I think the idiomatic Go way to do this would be as a method, which the caller can test via type assertion.

**Comment by cyriltovena** [comment]
> I found it rather hard to follow this code, because it's really doing two different things, and the second commit undoes part of the first commit without explanation. I think I understand what is intended now, after some Slack conversation.

The first commit fixed the `ContentLength` but then I wasn't able to get the `bytes.buffer` like I wanted because see below.

> The addition of `nopCloser` should be motivated; you declare a similar thing to be "useless" in your description.

I had to introduce `nopCloser` because if the body is not a `ReaderCloser` the NewRequest will do the wrapping with io.NopCloser, the problem is that this one is private and it's impossible to get the body without reflection.



See the full NewRequest code. https://golang.org/src/net/http/request.go?s=28366:28467#857

```
	rc, ok := body.(io.ReadCloser)
	if !ok && body != nil {
		rc = ioutil.NopCloser(body)
	}
```

I missed that at first, this is why it's another commit and that's what motivates this new struct.

**Comment by bboreham** [comment]
Please rebase to create a set of commits where each commit is explained by its description and does not undo part of an earlier commit.

File: httpgrpc/server/server.go [hunk 2]
```diff
@@ -37,16 +38,35 @@ func NewServer(handler http.Handler) *Server {
...
+
+func (nopCloser) Close() error { return nil }
+
+func nopCloserBuffer(b *bytes.Buffer) io.ReadCloser {
+	return nopCloser{b}
+}
+
+// BytesBufferFromReader attempts to cast a `io.ReadCloser` as a `bytes.Buffer` to avoid reading and copying one that already exists.
+// In case of failure it will return nil and a false boolean.
+func BytesBufferFromReader(r io.ReadCloser) (*bytes.Buffer, bool) {
```
Comment (bboreham) [inline httpgrpc/server/server.go:53]: I think the idiomatic Go way to do this would be as a method, which the caller can test via type assertion.

Comment (cyriltovena) [reply httpgrpc/server/server.go:53]: In my opinion, I think this is simpler, what you described means I need to know about the type and the method. 

I'll do the change anyway, it'll save time for both of us.

**Comment by cyriltovena** [comment]
Closing in favor of:

- https://github.com/weaveworks/common/pull/211
- https://github.com/weaveworks/common/pull/210

File: httpgrpc/server/server.go (hunk 2 above)
Comment (bboreham) [reply httpgrpc/server/server.go:53]: Just the method.  Example: https://github.com/grpc/grpc-go/blob/4be647f/rpc_util.go#L731-L733
